src/packet_analysis/protocol/teredo/Teredo.cc

```diff
@@ -270,10 +270,62 @@ bool TeredoAnalyzer::DetectProtocol(size_t len, const uint8_t* data, Packet* pac
 	if ( ! BifConst::Tunnel::enable_teredo )
 		return false;
 
-	if ( ! pattern_re->Match(data, len) )
+	if ( len < 40 )
 		return false;
 
-	return true;
+	if ( pattern_re->Match(data, len) )
+		return true;
+
+	// If the above regex failed, a Teredo packet *must* start with a zero in the first
+	// byte. Dump anything otherwise.
+	if ( data[0] != 0x00 )
+		return false;
+
+	uint16_t val = data[1];
+
+	if ( val == 1 )
```

Comment:
Looks different and more commented, but the basis for this was the Wireshark source you mentioned I guess? Should we put a link in the commit message or even inline as a reference/credit? Any rules around that?

https://gitlab.com/wireshark/wireshark/-/blob/master/epan/dissectors/packet-teredo.c#L237-302

Reply:
Hmm, taking a step back, `Teredo::DoParse()` seems to do what `DetectProtocol()` now does as well including weird logging.

So we have the parsing logic twice now which makes me think the naive initial patch to trigger the violation only after confirmation just as good/bad? I can see either.

What plays into this all is that for packet analyzer we don't have an easy way to disable / remove the analyzer from a session like we can do for protocol analyzer. So `DetectProtocol()` will run for any following packets of a session, too.

Reply:
> So we have the parsing logic twice now which makes me think the naive initial patch to trigger the violation only after confirmation just as good/bad? I can see either.

I can as well. My concern with just adding the confirmation checks is that we're still going to be calling `AnalyzePacket` for a bunch of packets that should have been ignored. Do you think it'd be enough to add a very-naive check for a packet starting with either `0x00` or in the range of `0x60`-`0x6f`? That should kick out a lot of packets and not duplicate all of the work in `DoParse`.

Maybe just add both checks?

Reply:
> Do you think it'd be enough to add a very-naive check for a packet starting with either 0x00 or in the range of 0x60-0x6f?

You mean in DetectProtocol() and removing the regex then? I have unfortunately not a good intuition :-(

I also realize that in AnalzyerPacket() there's the short-circuiting for when a violation happened and that will stop being effective if we never call `AnalyzerViolation()`
```
	else if ( AnalyzerViolated(packet->session) )
		return false;
```

I'd be also fine with the duplication as is if that filters out most. Hopefully the other tunnels aren't as trigger-happy :-/

Reply:
> You mean in DetectProtocol() and removing the regex then? 

I think if I check for those two bytes, even before doing the regex, would cut out most things faster than waiting for the regex first.

> I'd be also fine with the duplication as is if that filters out most. Hopefully the other tunnels aren't as trigger-happy

Teredo and AYIYA are the only analyzers (tunnels or not) that implements DetectProtocol, because they were the only ones doing DPD before the tunnel analyzers got moved to be packet analyzers. AYIYA's uses some magic numbers in the first three bytes, but it's not nearly as broad as Teredo.

Reply:
Let me resolve that thread, I'm also okay without the reference and think I understood what you intend with the fast check.

Only last thought would to extend the pre-check with `(data[0] == 0x00) && (data[1] == 0x00 || data[1] = 0x01)` open-coding the other regex part.